compiler/rustc_codegen_ssa/src/mir/place.rs

```diff
@@ -387,13 +388,21 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
                     let niche_llty = bx.cx().immediate_backend_type(niche.layout);
                     let niche_value = variant_index.as_u32() - niche_variants.start().as_u32();
                     let niche_value = (niche_value as u128).wrapping_add(niche_start);
```

Comment:
```suggestion
                    // We are supposed to compute `niche_value.wrapping_add(niche_start)` wrapping
                    // around the `niche`'s type. Since we know that type to be unsigned,
                    // the easiest way to do that is to do wrapping arithmetic on `u128` and then
                    // masking off any extra bits that occur because we did the arithmetic with too many bits.
                    let niche_value = (niche_value as u128).wrapping_add(niche_start);
                    let niche_value = niche_value & niche.layout.size.unsigned_int_max();
```

Reply:
Turns out it can be signed, but the logic should be right even for that case.